README.md
<h1 align="center">
  <br>
  <a href="https://github.com/teob97/T-RayX/"><img src="logo/T-RayX.png" alt="T-RayX" width="300" height="300"></a>
  <br>
  T-RayX
  <br>
</h1>

<h2 align="center">A <a href="https://nim-lang.org/" target="_blank">NIM</a> Raytracing Library.</h2>

<p align="center">
  <a href='https://github.com/teob97/T-RayX/releases'>
  <img src='https://img.shields.io/github/v/release/teob97/T-RayX?color=%23FDD835&label=version&style=for-the-badge'>
  </a>
  <a href='https://github.com/teob97/T-RayX/blob/main/LICENSE'>
  <img src='https://img.shields.io/github/license/teob97/T-RayX?style=for-the-badge'>
  </a>
  <a href="https://www.paypal.me/EGatti619">
  <img src='https://img.shields.io/badge/Donate-PayPal-DE00A5.svg?style=for-the-badge'>
  </a>
</p>

---

## :t-rex:  Overview
T-RayX: a Nim library aimed to generate a photorealistic image.

The project has been developed during the course [*Numerical techniques for photorealistic image generation*](https://www.unimi.it/en/education/degree-programme-courses/2022/numerical-tecniques-photorealistic-image-generation) held by Prof. [Maurizio Tomasi](https://github.com/ziotom78) at Università degli Studi di Milano (A.Y. 2021/2022)

The main functionality of this library is to generate photorealistic images from input files that describe a scene. (See more details [here](#small_orange_diamond-render)).
With T-RayX you are also able to convert PFM files to PNG using the [pfm2png](#pfm2png) command.

## :desktop_computer:  System Requirements

T-RayX works on Linux, MacOSX and Windows (not so sure) machine. 

For a proper use of the library you need:
- [Nim](https://nim-lang.org/) version required: 1.6.4
- [Nimble](https://github.com/nim-lang/nimble) package manager
- [simplepng](https://github.com/jrenner/nim-simplepng): use ```nimble install simplepng``` to install it.
- [docopt](https://github.com/docopt/docopt.nim): use ```nimble install docopt``` to install it.
- [ffmpeg](https://ffmpeg.org/) and [GNU parallel](https://www.gnu.org/software/parallel/) just for the animations (see _scripts_ folder).

## :wrench: Download and build

You can download the latest stable release [here](https://github.com/teob97/T-RayX/releases), and unpack it
   ``` bash
   tar -xvf /path/to/tar #or zip file -C /path/to/your/directory
   ```
or if you want, you can clone this repository
   ``` bash
   git clone https://github.com/teob97/T-RayX.git
   ```
   
To generate the executable file, use:
```bash
nimble build -d:release
```
It is also possible to install T-RayX using nimble package menager:

```bash
nimble intall trayx
```
In this case the program is automaticaly compiled and it is possible to use it via terminal simply with `trayx` without `./`

## :rocket:  Usage

You can run the following command to visualize through the CLI all the possible procedures:
```bash
./trayx --help
```

### :small_orange_diamond: render

To run the render, use:

```bash
./trayx render <SCENE_FILE.txt> <width> <height> [options]
```
To see all the possible options use `./trayx --help`.

The input file SCENE_FILE.txt must contain the detailed description of the scene that has to be rendered. See [the next section](#page_facing_up-input-file-format)

### :small_orange_diamond: pfm2png

Convert pfm file in png image using:

```bash
./trayx pfm2png <file.pfm> <alpha> <gamma> <output.png>
```

It is necessary to set specific values for alpha and gamma parameters.

### :small_orange_diamond: demo

To run the demo, use:

```bash
./trayx demo
```
This will produce the following 960x540 image:
<p float="center">
  <img src="examples/demo.png" width="600" />
</p>

## :page_facing_up: Input file format

To describe the scene that has to be rendered it is necessary to create a .txt file and use a specific syntax. Below are all the keywords that can be used within the input file:

| **FIGURE**  | **BRDF** | **PIGMENT** | **TRANSFORMATION** | **CAMERA_TYPE** | **_Extras_** |
|---------------|------------|---------------|----------------------|--------------|--------------|
| plane         | diffuse    | uniform       | identity             | orthogonal      | new          |
| sphere        | specular   | checkered     | translation          | perspective   | float        |
| aabox         |            | image         | rotation_x           |              | light        |
| cylinder      |            |               | rotation_y           |              | material     |
|               |            |               | rotation_z           |              | camera       |
|               |            |               | scaling              |              |              |

First thing first it is necessary to define a camera, use the following syntax:

```
camera(CAMERA_TYPE, TRANSFORMATION, aspectRatio:float, distance:float)
```
where `aspectRatio` is a float representing the ration between the width and the height (expressed in # of pixel) of the image; `distance` is a float representing the distance of the camera from the screen.

To define a space vector or a color it is necessary to specify three values:
```
[x:float, y:float, z:float] # This is a vector representing a 3D position or a direction
<r:float, g:float, b:float> # This is a vector representing a RGB color
```
The first vectors are used to define translation. Here are the ways to define the various possible transformations:
```
identity # Use this if you don't want a transformation
translation([x,y,z])
scaling([x,y,z])
rotation_x(rotation_angle:float) # Define a rotation around the x axis
rotation_y(rotation_angle:float) # Define a rotation around the y axis
rotation_z(rotation_angle:float) # Define a rotation around the z axis
```
To define a pigment use the RGB vectors:
```
uniform(<r,g,b>)
chekered(<r1,g1,b1>, <r2,g2,b2>, n_step:int)
image(image_PATH:string)
```
To define the material of the shapes that will be rendered it is necessary to use the correct BDRF:
```
material NAME(diffuse(PIGMENT), uniform(<r,g,b>)) # Diffuse brdf, the second pigment represent the emitted radiance.
material NAME(specular(PIGMENT), uniform(<0,0,0>)) # Specural brdf
```
To define a shape use this syntax:
```
FIGURE (MATERIAL, TRANSFORMATION)
```
In _pointlight_ render to define a light source use:
```
light ([x,y,z], <r,g,b>, radius:float)
```
It is also possible to define float variables:
```
float NAME(value:float)
```
You can find many examples in the [examples](./examples) folder!

## :book: Documentation

A complete documentation with all the procedures and the data structures is aviable at this [link](https://teob97.github.io/T-RayX/)

## :police_car: License

The code is released under a GPL-3.0 license. See the file [LICENSE](./LICENSE)

## :sunglasses: Examples

### Example 1 (pfm2png)

Run:

```bash
$ ./trayx pfm2png examples/pfm2png/lawn.pbm 0.6 1.45 examples/pfm2png/lawn_a0.6-gamma1.45.png
```

in order to create the following image:

<p float="center">
  <img src="examples/pfm2png/lawn_a0.6-gamma1.45.png" width="300" />
</p>

It is possible to tune the parameters alpha and gamma.

![](examples/pfm2png/lawn_a0.3-gamma1.45.png)  |  ![](examples/pfm2png/lawn_a0.6-gamma1.45.png) | ![](examples/pfm2png/lawn_a0.9-gamma1.45.png) 
:--:|:--:|:--:|
`alpha = 0.3` | `alpha = 0.6`  |  `alpha = 0.9`

### Example 2 (render)

Use the following command to produce the scene described [here](examples/render/example1.txt):

```bash
./trayx render examples/render/example1.txt 400 400 --output=examples/render/example1.png --luminosity=0.4 --gamma=2 --samplePerPixel=9

```

Use the following command to produce the scene described [here](examples/render/example2.txt):

```bash
./trayx render examples/render/example2.txt 400 400 --output=examples/render/example2.png --luminosity=0.2 --gamma=2 --maxDepth=3 --samplePerPixel=9
```

Use the following command to produce the scene described [here](examples/render/example3.txt):

```bash
./trayx render examples/render/example3.txt 400 400 --output=examples/render/example3.png --samplePerPixel=9
```

![](examples/render/example1.png)  |  ![](examples/render/example2.png) | ![](examples/render/example3.png) 
:--:|:--:|:--:|
`example 1` | `example 2`  |  `example 3`

You can also try to generate more artistic images. For the following image, we took inspiration from an artistic work by [Jeff Koons](http://www.jeffkoons.com/artwork/celebration/balloon-dog-0).

<p float="center">
  <img src="examples/render/example4.png" width="600" />
</p>



CHANGELOG.md
# Version 1.0.1

- Fixed output bug [#9](https://github.com/teob97/T-RayX/pull/10)
- Now it is possible to install T-RayX using nimble package manager.

# Version 1.0.0

- Fixed main and CLI [#8](https://github.com/teob97/T-RayX/pull/8)
- Added the possibility to override a variable (use --defineFloat in the CLI) previously defined in the input .txt file.
- Now you can correctly render AABox and Cylinder using the PointLight tracer.
- **Major changes:** now it is possible to use `render` command through the CLI. [#7](https://github.com/teob97/T-RayX/pull/7)
- Added a parser and a lexer in scenefiles.nim so that a .txt file can be used as input. [#7](https://github.com/teob97/T-RayX/pull/7)

# Version 0.2.0

- Added Point Light Tracer [#6](https://github.com/teob97/T-RayX/pull/6)
- Added Antialiasing [#6](https://github.com/teob97/T-RayX/pull/6)
- Added the possibility to deploy differents shapes: Spheres, Axis Aligned Boxes and Cylinders [#6](https://github.com/teob97/T-RayX/pull/6)
- Added Path Tracer [#6](https://github.com/teob97/T-RayX/pull/6)
- **Major changes:** now it is possible to choose between `pfm2pdf` and `demo` command through a nice CLI. [#5](https://github.com/teob97/T-RayX/pull/5)
- Fix an issue with the vertical order of the images [#4](https://github.com/teob97/T-RayX/pull/4)

# Version 0.1.0

- First release of the code


src/trayx.nim
#[  
  T-RayX: a Nim ray tracing library
  Copyright (C) 2022 Matteo Baratto, Eleonora Gatti

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <https://www.gnu.org/licenses/>. 
]#

import trayx/[basictypes, cameras, pfm, ldr,  imagetracer, shapes, transformation, geometry, materials, renderer, scenefiles, pcg]
import std/[strutils, strformat, streams, times, monotimes, options, os]
import docopt
when compileOption("profiler"):
  import nimprof

let doc = """
T-RayX: a Nim Raytracing Library. 
https://github.com/teob97/T-RayX

Usage:
  ./trayx render <SCENE_FILE.txt> <width> <height> [options]
  ./trayx pfm2png <INPUT_FILE.pfm> <alpha> <gamma> <OUTPUT.png>
  ./trayx demo

Options:
  --renderer=<type>             Renderer's type: onoff, flat, pathtracing, pointlight. [default: pathtracing]
  --output=<output-file>        Output file.png
  --numberOfRays=<nRay>         Number of rays departing from each surface point (pathtracing). [default: 10]
  --maxDepth=<depth>            Maximum allowed number of ray reflection (pathtracing). [default: 2]
  --russian=<RussianLimit>      Depth beyond which the Russian Roulette is triggered. [default: 3]
  --alpha=<alpha>               Normalization parameter used during tone mapping. [default: 1.0]
  --gamma=<gamma>               Correction due to the monitor non-linear response. [default: 1.0]
  --initState=<seed>            Initial seed (positive int) for the random number generator. [default: 42]
  --initSeq=<seq-seed>          Identifier (positive int) of the sequence produced by the random number generator. [default: 97]
  --samplePerPixel=<n_sample>   Number of samples per pixel (must be a perfect square, e.g. 2,4,16...). [default: 0]
  --luminosity=<lum>            Average luminosity of the image (positive float).
  --defineFloat=<var:value>     Used to declare a new float variable. Use '/' to define multiple variables.
Other:
  -h --help                     Show this screen.
  --version                     Show version.
"""

const versionStr = (staticExec "git describe --tags HEAD").split('-')[0]
let args = docopt(doc, version = versionStr)

#*********************************** PFM2PNG ***********************************

proc pfm2png*() =
  ## Procedure to convert a .pfm file into a .png file.
  ## Use: ./trayx pfm2png <INPUT_FILE.pfm> <alpha> <gamma> <OUTPUT.png>
  var 
    impf = openFileStream($args["<INPUT_FILE.pfm>"])
    img : HdrImage = readPfmImage(impf)
  impf.close()
  echo ("File "&($args["<INPUT_FILE.pfm>"])&" has been read from disk.")
  img.normalize_image(factor = parseFloat($args["<alpha>"]))
  img.clamp_image()
  var outf = newFileStream($args["<OUTPUT.png>"], fmWrite)
  img.writeLdrImage(name = $args["<OUTPUT.png>"], gamma = parseFloat($args["<gamma>"]))
  outf.close()
  echo (fmt"File "&($args["<OUTPUT.png>"])&" has been written to disk.")

#*********************************** DEMO ***********************************

proc demo*()=
  ## A demo procedure that generate an image output/demo.png using the pathtracing algoritm.
  ## Use: ./trayx demo
  ## Elements in the image: 
  ## diffusive sphere, 
  ## reflective sphere,
  ## checkered plane as ground
  ## and a luminous sphere as sky.
  let
    width  : int = 960
    height : int = 540
    ratio : float = width/height
  var 
    transf : Transformation = translation(newVec(-1.0, 0.0, 1.0))
    world : World
    strm : FileStream = newFileStream("output/demo.pfm", fmWrite)
    tracer : ImageTracer = newImageTracer(newHdrImage(width, height), newPerspectiveCamera(1, ratio, transf))      
  let
    sky_material: Material = newMaterial(brdf = newDiffuseBRDF(newUniformPigment(newColor(0, 0, 0))), emitted_radiance = newUniformPigment(newColor(1.0, 0.9, 0.5)))
    ground_material: Material = newMaterial(brdf = newDiffuseBRDF(pigment = newCheckeredPigment(color1 = newColor(0.3, 0.5, 0.1), color2 = newColor(0.1, 0.2, 0.5))))
    sphere_material: Material = newMaterial(brdf = newDiffuseBRDF(pigment = newUniformPigment(newColor(0.3, 0.4, 0.8))))
    mirror_material: Material = newMaterial(brdf = newSpecularBRDF(pigment = newUniformPigment(color = newColor(0.6, 0.2, 0.3))))
  # Add all the shapes in world
  world.shapes.add(newSphere(material=sky_material, transformation=scaling(newVec(200, 200, 200)) * translation(newVec(0, 0, 0.4))))
  world.shapes.add(newPlane(material=ground_material))
  world.shapes.add(newSphere(material=sphere_material, transformation=translation(newVec(0, 0, 1))))
  world.shapes.add(newSphere(material=mirror_material, transformation=translation(newVec(1, 2.5, 0))))
  #Initiallize the render and fire the rays
  let renderer : Renderer = newPathTracer(world)
  tracer.fireAllRays(renderer)
  tracer.image.writePfmImage(strm)
  let path : string = "output/demo.png"
  tracer.image.normalizeImage(factor = 1.0)
  tracer.image.clampImage()
  tracer.image.writeLdrImage(path)
  echo("Image saved in "&path)

#*************************************RENDER*************************************************

proc build_variable_table(cliInput : string) : Table[string, float] =
  ## Build a table (dictionary) with the variables defined throught the CLI,
  ## This dictionary will be used to override the values in the parser.
  for elements in cliInput.split('/'):
    var nameAndValues = elements.split(':')
    if nameAndValues.len() != 2:
      raise newException(IOError, "Invalid number of parameters in --defineFloat. Use var1:value1/var2:value2")
    var floatValue : float = nameAndValues[1].parseFloat()
    result[nameAndValues[0]]=floatValue

proc render*() =
  ## Main procedure to renderize an image using a specific alghoritm.
  ## Use: ./trayx render <SCENE_FILE.txt> <width> <height> [options]
  ## To see the aviable options use: ./trayx --helps 
  
  var variables = initTable[string, float]()
  if args["--defineFloat"]:
    variables = build_variable_table($args["--defineFloat"])
  
  # Define input stream and parse the scene
  var 
    file_stream : FileStream = newFileStream($args["<SCENE_FILE.txt>"], fmRead)
    input_stream : InputStream = newInputStream(file_stream)
    img_scene : Scene = parseScene(input_stream, variables)
  file_stream.close()
  
  # Define renderer
  var  
    renderer : Renderer
  if args["--renderer"]:
    case $args["--renderer"]:
      of "onoff":
        renderer = newOnOffRenderer(img_scene.world)
      of "flat":
        renderer = newFlatRenderer(img_scene.world)
      of "pointlight":
        renderer = newPointLightRenderer(img_scene.world)
      of "pathtracing":
        renderer = newPathTracer(world = img_scene.world,
                                pcg = newPCG(init_state = parseUInt($args["--initState"]), init_seq = parseUInt($args["--initSeq"])),
                                num_of_rays = parseInt($args["--numberOfRays"]),
                                max_depth = parseInt($args["--maxDepth"]),
                                russian_roulette_limit = parseInt($args["--russian"])
                                )
      else:
        raise newException(IOError, "Invalid type of renderer.")

  # Define image parameters and image tracer
  let
    width  : int = parseInt($args["<width>"])
    height : int = parseInt($args["<height>"])
  
  var 
    tracer : ImageTracer = newImageTracer(image = newHdrImage(width, height), 
                                            camera = img_scene.camera.get(),
                                            samples_per_side = parseInt($args["--samplePerPixel"]),
                                            pcg = newPCG()
                                            )

  # Fire all the rays and solve the render equation for every ray
  tracer.fireAllRays(renderer)
  
  # Save the final image
  let
    path_ldr : string = "output/img_"&(now().format("yyyy-MM-dd'T'HH:mm:ss"))&".png"
    path_pfm : string = "output/img_"&(now().format("yyyy-MM-dd'T'HH:mm:ss"))&".pfm"
    out_strm : FileStream = newFileStream(path_pfm, fmWrite)

  tracer.image.writePfmImage(out_strm)

  if args["--luminosity"]:
    tracer.image.normalizeImage(factor = parseFloat($args["--alpha"]), luminosity = some(parseFloat($args["--luminosity"])))
  else:
    tracer.image.normalizeImage(factor = parseFloat($args["--alpha"]))
  
  tracer.image.clampImage()

  if args["--output"]:
    tracer.image.writeLdrImage($args["--output"], gamma = parseFloat($args["--gamma"]))
  else:
    tracer.image.writeLdrImage(path_ldr, gamma = parseFloat($args["--gamma"]))

#*********************************** MAIN ***********************************

when isMainModule:

  if not dirExists("output"):
    createDir("output")

  if args["pfm2png"]:
    let t1 = getMonoTime()
    pfm2png()
    let t2 = getMonoTime()
    echo("Execution time: ", t2 - t1)  

  if args["demo"]:
    let t1 = getMonoTime()
    demo()
    let t2 = getMonoTime()
    echo("Execution time: ", t2 - t1)

  if args["render"]:
    let t1 = getMonoTime()
    render()
    let t2 = getMonoTime()
    echo("Execution time: ", t2 - t1)